src/b64.ml

```diff
@@ -16,77 +19,144 @@
+
+let length_alphabet { emap; _ } = Array.length emap
+
+let default_alphabet = make_alphabet "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/"
+let uri_safe_alphabet = make_alphabet "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789-_"
+
+let unsafe_set_be_uint16 =
+  if Sys.big_endian
+  then fun t off v -> unsafe_set_uint16 t off v
+  else fun t off v -> unsafe_set_uint16 t off (swap16 v)
+
+exception Out_of_bounds
+
+let get_uint8 t off =
+  if off < 0 || off >= String.length t then raise Out_of_bounds ;
+  unsafe_get_uint8 t off
+
+let padding = int_of_char '='
+
+let encode pad { emap; _ } input =
+  let n = String.length input in
+  let n' = n // 3 * 4 in
+  let res = Bytes.create n' in
+
+  let emap i = Array.unsafe_get emap i in
+
+  let emit b1 b2 b3 i =
+    unsafe_set_be_uint16 res i
+      ((emap (b1 lsr 2 land 0x3f) lsl 8)
+       lor (emap ((b1 lsl 4) lor (b2 lsr 4) land 0x3f))) ;
+    unsafe_set_be_uint16 res (i + 2)
+      ((emap ((b2 lsl 2) lor (b3 lsr 6) land 0x3f) lsl 8)
+       lor (emap (b3 land 0x3f))) in
+
+  let rec enc j i =
+    if i = n then ()
+    else if i = n - 1 then emit (unsafe_get_uint8 input i) 0 0 j
+    else if i = n - 2 then emit (unsafe_get_uint8 input i) (unsafe_get_uint8 input (i + 1)) 0 j
+    else
+    (emit
+       (unsafe_get_uint8 input i)
+       (unsafe_get_uint8 input (i + 1))
+       (unsafe_get_uint8 input (i + 2))
+       j ;
+     enc (j + 4) (i + 3)) in
+
+  let rec unsafe_fix = function
+  | 0 -> ()
+  | i -> unsafe_set_uint8 res (n' - i) padding ; unsafe_fix (i - 1) in
+
+  enc 0 0 ;
+
+  if pad
+  then begin unsafe_fix ((3 - n mod 3) mod 3) ; Bytes.unsafe_to_string res end
+  else Bytes.sub_string res 0 (n / 3 * 4)
+
+let encode ?(pad = true) ?(alphabet = default_alphabet) input = encode pad alphabet input
+
+let error_msgf fmt = Format.ksprintf (fun err -> Error (`Msg err)) fmt
+
+let decode_result { dmap; _ } input =
+  let n = String.length input in
+  let n' = (n / 4) * 3 in
+  let res = Bytes.create n' in
+
+  let emit a b c d i =
+    let x = (a lsl 18) lor (b lsl 12) lor (c lsl 6) lor d in
+    unsafe_set_be_uint16 res i (x lsr 8) ;
+    unsafe_set_uint8 res (i + 2) (x land 0xff) in
+
+  let dmap i =
+    let x = Array.unsafe_get dmap i in
+    if x = none then raise Not_found else x in
+
+  let rec dec j i =
+    if i = n then Some 0
+    else begin
+      let a = dmap (get_uint8 input i) in
+      let b = dmap (get_uint8 input (i + 1)) in
+      let (d, pad) =
+        let x = get_uint8 input (i + 3) in
+        try (dmap x, 0) with Not_found when x = padding -> (0, 1) in
+      let (c, pad) =
+        let x = get_uint8 input (i + 2) in
+        try (dmap x, pad) with Not_found when x = padding && pad = 1 -> (0, 2) in
+
+      emit a b c d j ;
+      if pad = 0
+      then dec (j + 3) (i + 4)
+      else if i + 4 <> n then None
+      else Some pad end in
+
+  match dec 0 0 with
+  | Some pad -> Ok (Bytes.sub_string res 0 (n' - pad))
```

Comment:
After all the tricks with unsafe accesses above, its a pity that we force a string copy just to remove some padding. Do you think it's worth having a slightly lower level return API of (string, length) so the caller can choose to take a substring themselves?  Similar to `Astring.String.Sub`

Reply:
I wondered about this in the case of
```ocaml
| Some 0 -> Ok (Bytes.unsafe_to_string res)
```

Reply:
Hmmhmm, it's probably ok to do that. However, I would like to know feedbacks of everyone about that before to do something.